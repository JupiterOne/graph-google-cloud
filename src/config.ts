import {
  IntegrationExecutionContext,
  IntegrationLogger,
} from '@jupiterone/integration-sdk-core';
import { ServiceUsageClient } from './steps/service-usage/client';
import { IntegrationConfig, SerializedIntegrationConfig } from './types';
import { deserializeIntegrationConfig } from './utils/integrationConfig';
import { google } from 'googleapis';
import { handleApiClientError } from './google-cloud/client';
import { ServiceUsageListFilter } from './google-cloud/types';

export async function executeTestRequest(
  config: IntegrationConfig,
  logger: IntegrationLogger,
) {
  try {
    const client = google.serviceusage({ version: 'v1', retry: false });
    const googleClient = new ServiceUsageClient({ config }, logger);
    const auth = await googleClient.getAuthenticatedServiceClient();

    await client.services.list({
      parent: `projects/${config.serviceAccountKeyConfig.project_id}`,
      pageSize: 200,
      auth: auth,
      filter: ServiceUsageListFilter.ENABLED,
    });
  } catch (err) {
    throw handleApiClientError(err);
  }
}

export async function validateInvocation(
  context: IntegrationExecutionContext<SerializedIntegrationConfig>,
) {
  const { instance, logger } = context;
  const { config: serializedIntegrationConfig } = instance;
  const config = (context.instance.config = deserializeIntegrationConfig(
    serializedIntegrationConfig,
  ));

  await executeTestRequest(config, logger);
}
