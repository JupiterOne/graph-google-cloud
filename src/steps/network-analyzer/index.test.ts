import {
  Recording,
  StepTestConfig,
  createMockStepExecutionContext,
  executeStepWithDependencies,
} from '@jupiterone/integration-sdk-testing';
import {
  STEP_NETWORK_ANALYZER_CONNECTIVITY_TEST,
  STEP_NETWORK_INTELLIGENCE_CENTER,
  STEP_VPN_GATEWAY,
  STEP_VPN_GATEWAY_TUNNEL,
  VPN_GATEWAY_CLASS,
  VPN_GATEWAY_TUNNEL_CLASS,
  VPN_GATEWAY_TUNNEL_TYPE,
  VPN_GATEWAY_TYPE,
} from './constants';
import { integrationConfig } from '../../../test/config';
import {
  setupGoogleCloudRecording,
  getMatchRequestsBy,
} from '../../../test/recording';
import { IntegrationConfig, invocationConfig } from '../..';
import { fetchVpnGateway, fetchVpnGatewayTunnel } from '.';

const tempNewAccountConfig = {
  ...integrationConfig,
  serviceAccountKeyFile: integrationConfig.serviceAccountKeyFile.replace(
    'j1-gc-integration-dev-v2',
    'j1-gc-integration-dev-v3',
  ),
  serviceAccountKeyConfig: {
    ...integrationConfig.serviceAccountKeyConfig,
    project_id: 'j1-gc-integration-dev-v3',
  },
};

describe(`networkAnalyzer#${STEP_NETWORK_INTELLIGENCE_CENTER}`, () => {
  let recording: Recording;
  afterEach(async () => {
    await recording.stop();
  });

  jest.setTimeout(45000);

  test(STEP_NETWORK_INTELLIGENCE_CENTER, async () => {
    recording = setupGoogleCloudRecording({
      name: STEP_NETWORK_INTELLIGENCE_CENTER,
      directory: __dirname,
      options: {
        matchRequestsBy: getMatchRequestsBy(tempNewAccountConfig),
      },
    });

    const stepTestConfig: StepTestConfig = {
      stepId: STEP_NETWORK_INTELLIGENCE_CENTER,
      instanceConfig: tempNewAccountConfig,
      invocationConfig: invocationConfig as any,
    };

    const result = await executeStepWithDependencies(stepTestConfig);
    expect(result).toMatchStepMetadata(stepTestConfig);
  });
});

describe(`networkAnalyzer#${STEP_NETWORK_ANALYZER_CONNECTIVITY_TEST}`, () => {
  let recording: Recording;
  afterEach(async () => {
    await recording.stop();
  });

  jest.setTimeout(45000);

  test(STEP_NETWORK_ANALYZER_CONNECTIVITY_TEST, async () => {
    recording = setupGoogleCloudRecording({
      name: STEP_NETWORK_ANALYZER_CONNECTIVITY_TEST,
      directory: __dirname,
      options: {
        matchRequestsBy: getMatchRequestsBy(tempNewAccountConfig),
      },
    });

    const stepTestConfig: StepTestConfig = {
      stepId: STEP_NETWORK_ANALYZER_CONNECTIVITY_TEST,
      instanceConfig: tempNewAccountConfig,
      invocationConfig: invocationConfig as any,
    };

    const result = await executeStepWithDependencies(stepTestConfig);
    expect(result).toMatchStepMetadata(stepTestConfig);
  });
});

describe(`networkAnalyzer#${STEP_VPN_GATEWAY}`, () => {
  let recording: Recording;

  beforeEach(() => {
    recording = setupGoogleCloudRecording({
      directory: __dirname,
      name: 'fechVpngateway',
    });
  });

  afterEach(async () => {
    await recording.stop();
  });

  test('should collect data', async () => {
    const context = createMockStepExecutionContext<IntegrationConfig>({
      instanceConfig: tempNewAccountConfig,
    });

    await fetchVpnGateway(context);

    expect({
      numCollectedEntities: context.jobState.collectedEntities.length,
      numCollectedRelationships: context.jobState.collectedRelationships.length,
      collectedEntities: context.jobState.collectedEntities,
      collectedRelationships: context.jobState.collectedRelationships,
      encounteredTypes: context.jobState.encounteredTypes,
    }).toMatchSnapshot();

    expect(
      context.jobState.collectedEntities.filter(
        (e) => e._type === VPN_GATEWAY_TYPE,
      ),
    ).toMatchGraphObjectSchema({
      _class: VPN_GATEWAY_CLASS,
      schema: {
        additionalProperties: false,
        properties: {
          _type: { const: 'google_cloud_vpn_gateway' },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },

          name: { type: 'string' },
          projectId: { type: 'string' },
          summary: { type: 'string' },
          public: { type: 'boolean' },
          kind: { type: 'string' },
          category: { type: 'array' },
          function: { type: 'array' },
        },
      },
    });
  }, 50000);
});

describe(`networkAnalyzer#${STEP_VPN_GATEWAY_TUNNEL}`, () => {
  let recording: Recording;

  beforeEach(() => {
    recording = setupGoogleCloudRecording({
      directory: __dirname,
      name: 'fetchVpnTunnel',
    });
  });

  afterEach(async () => {
    await recording.stop();
  });

  test('should collect data', async () => {
    const context = createMockStepExecutionContext<IntegrationConfig>({
      instanceConfig: tempNewAccountConfig,
    });

    await fetchVpnGatewayTunnel(context);

    expect({
      numCollectedEntities: context.jobState.collectedEntities.length,
      numCollectedRelationships: context.jobState.collectedRelationships.length,
      collectedEntities: context.jobState.collectedEntities,
      collectedRelationships: context.jobState.collectedRelationships,
      encounteredTypes: context.jobState.encounteredTypes,
    }).toMatchSnapshot();

    expect(
      context.jobState.collectedEntities.filter(
        (e) => e._type === VPN_GATEWAY_TUNNEL_TYPE,
      ),
    ).toMatchGraphObjectSchema({
      _class: VPN_GATEWAY_TUNNEL_CLASS,
      schema: {
        additionalProperties: false,
        properties: {
          _type: { const: 'google_cloud_vpn_tunnel' },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },

          name: { type: 'string' },
          projectId: { type: 'string' },
          summary: { type: 'string' },
          public: { type: 'boolean' },
          kind: { type: 'string' },
          category: { type: 'array' },
          function: { type: 'array' },
          vpnGatewayName: { type: 'string' },
        },
      },
    });
  }, 50000);
});
