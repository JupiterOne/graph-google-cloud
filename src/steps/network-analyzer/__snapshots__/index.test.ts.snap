// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`networkAnalyzer#fetch-cloud-vpn-gateway should collect data 1`] = `
{
  "collectedEntities": [
    {
      "_class": [
        "Gateway",
      ],
      "_key": "demo-vpn-gateway",
      "_rawData": [
        {
          "name": "default",
          "rawData": {
            "creationTimestamp": "2024-04-23T23:16:17.360-07:00",
            "gatewayIpVersion": "IPV4",
            "id": "5169371474615934302",
            "kind": "compute#vpnGateway",
            "labelFingerprint": "42WmSpB8rSM=",
            "name": "demo-vpn-gateway",
            "network": "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/global/networks/default",
            "region": "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/us-central1",
            "selfLink": "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/us-central1/vpnGateways/demo-vpn-gateway",
            "stackType": "IPV4_ONLY",
            "vpnInterfaces": [
              {
                "id": 0,
                "ipAddress": "34.157.99.170",
              },
              {
                "id": 1,
                "ipAddress": "35.220.75.52",
              },
            ],
          },
        },
      ],
      "_type": "google_cloud_vpn_gateway",
      "category": [
        "network",
      ],
      "createdOn": 1713939377360,
      "displayName": "demo-vpn-gateway",
      "function": [
        "routing",
        "remote-access-gateway",
      ],
      "id": "5169371474615934302",
      "kind": "compute#vpnGateway",
      "name": "demo-vpn-gateway",
      "projectId": "j1-gc-integration-dev-v3",
      "public": true,
      "summary": undefined,
    },
    {
      "_class": [
        "Gateway",
      ],
      "_key": "demo-vpn-gateway-2",
      "_rawData": [
        {
          "name": "default",
          "rawData": {
            "creationTimestamp": "2024-04-23T23:26:08.398-07:00",
            "gatewayIpVersion": "IPV4",
            "id": "506472523849079023",
            "kind": "compute#vpnGateway",
            "labelFingerprint": "42WmSpB8rSM=",
            "name": "demo-vpn-gateway-2",
            "network": "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/global/networks/default",
            "region": "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/us-east1",
            "selfLink": "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/us-east1/vpnGateways/demo-vpn-gateway-2",
            "stackType": "IPV4_ONLY",
            "vpnInterfaces": [
              {
                "id": 0,
                "ipAddress": "35.242.10.161",
              },
              {
                "id": 1,
                "ipAddress": "35.220.8.175",
              },
            ],
          },
        },
      ],
      "_type": "google_cloud_vpn_gateway",
      "category": [
        "network",
      ],
      "createdOn": 1713939968398,
      "displayName": "demo-vpn-gateway-2",
      "function": [
        "routing",
        "remote-access-gateway",
      ],
      "id": "506472523849079023",
      "kind": "compute#vpnGateway",
      "name": "demo-vpn-gateway-2",
      "projectId": "j1-gc-integration-dev-v3",
      "public": true,
      "summary": undefined,
    },
  ],
  "collectedRelationships": [],
  "encounteredTypes": [
    "google_cloud_vpn_gateway",
  ],
  "numCollectedEntities": 2,
  "numCollectedRelationships": 0,
}
`;

exports[`networkAnalyzer#fetch-cloud-vpn-gateway-tunnel should collect data 1`] = `
{
  "collectedEntities": [
    {
      "_class": [
        "Gateway",
      ],
      "_key": "demo-vpn-tunnel",
      "_rawData": [
        {
          "name": "default",
          "rawData": {
            "creationTimestamp": "2024-04-24T00:08:43.251-07:00",
            "description": "",
            "detailedStatus": "No incoming packets from peer",
            "id": "5333968030286549268",
            "ikeVersion": 2,
            "kind": "compute#vpnTunnel",
            "labelFingerprint": "42WmSpB8rSM=",
            "localTrafficSelector": [
              "0.0.0.0/0",
            ],
            "name": "demo-vpn-tunnel",
            "peerExternalGateway": "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/global/externalVpnGateways/demo-peer-gateway",
            "peerExternalGatewayInterface": 0,
            "peerIp": "12.120.12.33",
            "region": "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/us-central1",
            "remoteTrafficSelector": [
              "0.0.0.0/0",
            ],
            "router": "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/us-central1/routers/demo-router2",
            "selfLink": "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/us-central1/vpnTunnels/demo-vpn-tunnel",
            "sharedSecret": "*************",
            "sharedSecretHash": "4bcIUgeMVskvZeiR0f8c9n4s1AaP",
            "status": "NO_INCOMING_PACKETS",
            "vpnGateway": "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/us-central1/vpnGateways/demo-vpn-gateway",
            "vpnGatewayInterface": 1,
          },
        },
      ],
      "_type": "google_cloud_vpn_tunnel",
      "active": undefined,
      "category": [
        "network",
      ],
      "createdOn": 1713942523251,
      "description": "",
      "displayName": "demo-vpn-tunnel",
      "function": [
        "routing",
      ],
      "id": "5333968030286549268",
      "kind": "compute#vpnTunnel",
      "name": "demo-vpn-tunnel",
      "projectId": "j1-gc-integration-dev-v3",
      "public": true,
      "status": "NO_INCOMING_PACKETS",
      "summary": "",
      "vpnGatewayName": "demo-vpn-gateway",
    },
  ],
  "collectedRelationships": [],
  "encounteredTypes": [
    "google_cloud_vpn_tunnel",
  ],
  "numCollectedEntities": 1,
  "numCollectedRelationships": 0,
}
`;
